core/base/inc/ROOT/TExecutorCRTP.hxx

```diff
@@ -0,0 +1,352 @@
+
+   /// type definition in used in templated functions for not allowing mapping functions that return references.
+   /// The resulting vector elements must be assignable, references aren't.
+   template< class F, class... T>
+   using noReferenceCond = typename std::enable_if<"Function can't return a reference" && !(std::is_reference<typename std::result_of<F(T...)>::type>::value)>::type;
+
+   // Map
+   // These trailing return types allow for a compile time check of compatibility between function signatures and args,
+   // and a compile time check that the argument list implements a front() method (all STL sequence containers have it)
+   template<class F, class Cond = noReferenceCond<F>>
+   auto Map(F func, unsigned nTimes) -> std::vector<typename std::result_of<F()>::type>;
+   template<class F, class INTEGER, class Cond = noReferenceCond<F, INTEGER>>
+   auto Map(F func, ROOT::TSeq<INTEGER> args) -> std::vector<typename std::result_of<F(INTEGER)>::type>;
+   template<class F, class T, class Cond = noReferenceCond<F, T>>
+   auto Map(F func, std::initializer_list<T> args) -> std::vector<typename std::result_of<F(T)>::type>;
+   template<class F, class T, class Cond = noReferenceCond<F, T>>
+   auto Map(F func, std::vector<T> &args) -> std::vector<typename std::result_of<F(T)>::type>;
+   template<class F, class T, class Cond = noReferenceCond<F, T>>
+   auto Map(F func, const std::vector<T> &args) -> std::vector<typename std::result_of<F(T)>::type>;
+
+   // MapReduce
+   // The trailing return types also check at compile time whether redfunc is compatible with func,
+   // other than checking that func is compatible with the type of arguments.
+   // a static_assert check in TExecutorCRTP<subc>::Reduce is used to check that redfunc is compatible with the type returned by func
+   template<class F, class R, class Cond = noReferenceCond<F>>
+   auto MapReduce(F func, unsigned nTimes, R redfunc) -> typename std::result_of<F()>::type;
+   template<class F, class INTEGER, class R, class Cond = noReferenceCond<F, INTEGER>>
+   auto MapReduce(F func, ROOT::TSeq<INTEGER> args, R redfunc) -> typename std::result_of<F(INTEGER)>::type;
+   template<class F, class T, class R, class Cond = noReferenceCond<F, T>>
+   auto MapReduce(F func, std::initializer_list<T> args, R redfunc) -> typename std::result_of<F(T)>::type;
+   template<class F, class T, class R, class Cond = noReferenceCond<F, T>>
+   auto MapReduce(F func, const std::vector<T> &args, R redfunc) -> typename std::result_of<F(T)>::type;
+   template<class F, class T, class R, class Cond = noReferenceCond<F, T>>
+   auto MapReduce(F func, std::vector<T> &args, R redfunc) -> typename std::result_of<F(T)>::type;
+   template<class F, class T,class Cond = noReferenceCond<F, T>>
+   T* MapReduce(F func, std::vector<T*> &args);
+   template<class F, class T,class Cond = noReferenceCond<F, T>>
+   T* MapReduce(F func, const std::vector<T*> &args);
```

Comment:
I guess the argument should be a `std::vector<TObject*>` if we want to use TObject's `Merge`?

Reply:
I am not sure all TObjects implement Merge, but it will improve it, yes

Reply:
iiuc there are two requirements on `T`, one is that `T` inherits from `TObject`, the other that it implements the `Merge` method with `TObject`'s signature.

I think at compile-time one can easily check for the first but not the second.

Reply:
You can make sure that `T::Merge` is callable with the right arguments. I don't know whether its return type must be convertible to `bool` or whatever, but even that you can require. See e.g. https://github.com/root-project/root/blob/master/interpreter/cling/include/cling/Interpreter/RuntimePrintValue.h#L156:L158

Reply:
@Axel-Naumann doesn't work for polymorphic uses of `TObject*`, does it?

Reply:
If the `vector<T*>` has `TObject` as `T` then, no. But I don't think that's possible, because IIRC there's no `TObject::Merge`? So you should be able to check that `T::Merge` is callable?

Reply:
ah my bad, my whole line of reasoning in this thread is that `TObject::Merge` exists and is pure virtual. @xvallspl feel free to resolve, thank you @Axel-Naumann